atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -2551,6 +2561,163 @@ public void exceptionThrownWhenTooManyPostFilterIterationsOccur() {
+    public void metadataStoredForPutWithMetadata() {
+        Map<Cell, ChangeMetadata> internalMetadata = txManager.runTaskThrowOnConflict(txn -> {
+            txn.putWithMetadata(
+                    TABLE, ImmutableMap.of(TEST_CELL, ChangeMetadataAnnotatedValue.of(TEST_VALUE, TEST_METADATA)));
+            return unwrapSnapshotTransaction(txn).getChangeMetadataForWrites(TABLE);
+        });
+        assertThat(internalMetadata).containsExactlyEntriesOf(ImmutableMap.of(TEST_CELL, TEST_METADATA));
+    }
+
+    @Test
+    public void metadataRemovedForRegularPutAfterPutWithMetadata() {
+        Map<Cell, ChangeMetadata> internalMetadata = txManager.runTaskThrowOnConflict(txn -> {
+            txn.putWithMetadata(
+                    TABLE, ImmutableMap.of(TEST_CELL, ChangeMetadataAnnotatedValue.of(TEST_VALUE, TEST_METADATA)));
+            txn.put(TABLE, ImmutableMap.of(TEST_CELL, PtBytes.toBytes("feeling lucky")));
+            return unwrapSnapshotTransaction(txn).getChangeMetadataForWrites(TABLE);
+        });
+        assertThat(internalMetadata).isEmpty();
+    }
+
+    @Test
+    public void metadataStoredForDeleteWithMetadata() {
+        Map<Cell, ChangeMetadata> internalMetadata = txManager.runTaskThrowOnConflict(txn -> {
+            txn.deleteWithMetadata(TABLE, ImmutableMap.of(TEST_CELL, TEST_METADATA));
+            return unwrapSnapshotTransaction(txn).getChangeMetadataForWrites(TABLE);
+        });
+        assertThat(internalMetadata).containsExactlyEntriesOf(ImmutableMap.of(TEST_CELL, TEST_METADATA));
+    }
+
+    @Test
+    public void metadataRemovedForRegularDeleteAfterPutWithMetadata() {
+        Map<Cell, ChangeMetadata> internalMetadata = txManager.runTaskThrowOnConflict(txn -> {
+            txn.putWithMetadata(
+                    TABLE, ImmutableMap.of(TEST_CELL, ChangeMetadataAnnotatedValue.of(TEST_VALUE, TEST_METADATA)));
+            txn.delete(TABLE, ImmutableSet.of(TEST_CELL));
+            return unwrapSnapshotTransaction(txn).getChangeMetadataForWrites(TABLE);
+        });
+        assertThat(internalMetadata).isEmpty();
+    }
+
+    // We can save ourselves a separate concurrency test for deleteWithMetadata() since it just calls putInternal()
+    // with the empty byte array value.
+    @Test
+    public void storingMetadataIsAtomicWithLocalWrite() throws InterruptedException {
+        int numThreads = 20;
+        int numCells = 50;
+        long randomSeed = System.currentTimeMillis();
+
+        Random random = new Random(randomSeed);
+        SnapshotTransaction txn = unwrapSnapshotTransaction(txManager.createNewTransaction());
+
+        Set<Cell> cells = IntStream.range(0, numCells)
+                .mapToObj(cellIndex -> Cell.create(PtBytes.toBytes(cellIndex), PtBytes.toBytes(cellIndex)))
+                .collect(Collectors.toSet());
+
+        Map<Long, Map<Cell, ChangeMetadata>> metadataWrittenByThread =
+                concurrentlyWriteMetadataToCells(random, txn, cells, numThreads);
```

Comment:
The main magic is happening in this method call, which I refactored out so the test would be a bit more readable